tests/drivers/rpc_flash/src/main.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2020-2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/ztest.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/storage/flash_map.h>
+
+#define TEST_NODE		DT_NODELABEL(rpc_flash_controller)
+#define TEST_AREA_DEVICE	DEVICE_DT_GET(TEST_NODE)
+#define TEST_AREA_MAX 		DT_REG_SIZE(TEST_NODE)
+#define TEST_AREA_OFFSET	0x10000
+#define EXPECTED_SIZE		512
+#define CANARY			0xff
+
+static const struct device *const flash_dev = TEST_AREA_DEVICE;
+static struct flash_pages_info page_info;
+static uint8_t __aligned(4) expected[EXPECTED_SIZE];
+
+static void *flash_driver_setup(void)
+{
+	int rc;
+
+	zassert_true(device_is_ready(flash_dev));
+
+	const struct flash_parameters *flash_params =
+			flash_get_parameters(flash_dev);
+
+	flash_get_page_info_by_offs(flash_dev, TEST_AREA_OFFSET,
+				    &page_info);
+
+	/* Check if test region is not empty */
+	uint8_t buf[EXPECTED_SIZE];
+
+	rc = flash_read(flash_dev, TEST_AREA_OFFSET,
+			buf, EXPECTED_SIZE);
+	zassert_equal(rc, 0, "Cannot read flash");
+
+	/* Fill test buffer with random data */
+	for (int i = 0; i < EXPECTED_SIZE; i++) {
+		expected[i] = i;
+	}
+
+	/* Check if tested region fits in flash */
+	zassert_true((TEST_AREA_OFFSET + EXPECTED_SIZE) <= TEST_AREA_MAX,
+		     "Test area exceeds flash size");
+
+	/* Check if flash is cleared */
+	bool is_buf_clear = true;
+
+	for (off_t i = 0; i < EXPECTED_SIZE; i++) {
+		if (buf[i] != flash_params->erase_value) {
+			is_buf_clear = false;
+			break;
+		}
+	}
+
+	if (!is_buf_clear) {
+		/* Erase a nb of pages aligned to the EXPECTED_SIZE */
+		rc = flash_erase(flash_dev, page_info.start_offset,
+				(page_info.size *
+				((EXPECTED_SIZE + page_info.size - 1)
+				/ page_info.size)));
+
+		zassert_equal(rc, 0, "Flash memory not properly erased");
+	}
+
+	return NULL;
+}
+
+ZTEST(flash_driver, test_read_unaligned_address)
+{
+	int rc;
+	uint8_t buf[EXPECTED_SIZE];
+
+	rc = flash_write(flash_dev,
+			 page_info.start_offset,
+			 expected, EXPECTED_SIZE);
+	zassert_equal(rc, 0, "Cannot write to flash");
+
+	/* read buffer length*/
+	for (off_t len = 0; len < 25; len++) {
+		/* address offset */
+		for (off_t ad_o = 0; ad_o < 4; ad_o++) {
+			/* buffer offset; leave space for buffer guard */
+			for (off_t buf_o = 1; buf_o < 5; buf_o++) {
+				/* buffer overflow protection */
+				buf[buf_o - 1] = CANARY;
+				buf[buf_o + len] = CANARY;
+				memset(buf + buf_o, 0, len);
```

Comment:
IMHO you could just fill in the buf with CANARY; you know where your real data will start and end, and, assuming that there are no CANARies in `expected`, you can figure out whether boundaries of buffer have been crossed in either way.